### PR TITLE
build: document and cleanup Qt hacks

### PR DESCRIPTION
**Pull request by fanquake**
Follow up on removing `sed` usage in #19761. Also nice to revisit & cleanup before 5.15.x.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

File: depends/packages/miniupnpc.mk [hunk 1]
```diff
@@ -12,6 +12,10 @@ $(package)_build_opts_mingw32=-f Makefile.mingw
 $(package)_build_env+=CFLAGS="$($(package)_cflags) $($(package)_cppflags)" AR="$($(package)_ar)"
 endef
 
+# sed is used here to replace the host and version that will end up as part of
+# the miniupnpc User-Agent string, that will ultimately end up in bitcoind. i.e
+# before: User-Agent: Debian/10, UPnP/1.1, MiniUPnPc/2.0
+# after : User-Agent: x86_64-pc-linux-gnu, UPnP/1.1, MiniUPnPc/2.0.20180203
```
Comment (laanwj) [inline depends/packages/miniupnpc.mk:18]: unrelated to this change but: why are we inserting the architecture in here, seems like something not necessary to reveal

**Comment by laanwj** [comment]
Concept ACK it's good to document these things carefully.

File: depends/packages/miniupnpc.mk (hunk 1 above)
Comment (practicalswift) [reply depends/packages/miniupnpc.mk:18]: I agree. Why make life easier for an attacker? :)

File: depends/packages/miniupnpc.mk [hunk 2]
```diff
@@ -12,6 +12,10 @@ $(package)_build_opts_mingw32=-f Makefile.mingw
 $(package)_build_env+=CFLAGS="$($(package)_cflags) $($(package)_cppflags)" AR="$($(package)_ar)"
 endef
 
+# sed is used here to replace the host and version that will end up as part of
+# the miniupnpc User-Agent string, that will ultimately end up in bitcoind. i.e
```
Comment (MarcoFalke) [inline depends/packages/miniupnpc.mk:16]: Which symbol or variable name contains this ua string? Not the p2p ua string of the `version` message, right?

Comment (laanwj) [reply depends/packages/miniupnpc.mk:16]: no, not the P2P UA string, it's an UA string sent to the UPnP server (so broadcasted to the network you'[re connected to, if you have UPnP enabled)

File: depends/packages/miniupnpc.mk (hunk 1 above)
Comment (fanquake) [reply depends/packages/miniupnpc.mk:18]: My assumption is that it was being inserted to make depends more deterministic. However I think we can improve this, as there's no reason to reveal more of the version information either. Could leave the version as is `/2.0` and either drop the architecture, or insert something else?

Comment (laanwj) [reply depends/packages/miniupnpc.mk:18]: I guess `UPnP/1.1`, the version of the standard used, not the version of the software, is the only relevant thing here. The rest just aids exploiters if there's another remotely exploitable UPnP bug.

**Comment by fanquake** [comment]
I've rebased, dropped the miniupnpc change (will be dealt with separately), added a commit to drop `global_init_link_order` from the mac-qmake.conf and fixed the the TODO in the preprocessing doc commit.

**Comment by hebasto** [comment]
Not sure how the _"build: remove global_init_link_order from mac qt qmake.conf"_ (e1f2553e1148de9bd57818b40b5fe9da7ff5e246) commit could be tested?

**Comment by fanquake** [comment]
> Not sure how the "build: remove global_init_link_order from mac qt qmake.conf" (e1f2553) commit could be tested/reviewed?

I'd suggest diffing `bitcoin-qt` (for macOS) after building `qt` with and without `global_init_link_order` in mac-qmake.conf. 

You could also check the qt 5.9 source and verify that `global_init_link_order` is not used anywhere (the only two occurences are in mkspec files). As opposed to an option like `lib_version_first`, which you'll find multiple usages of, i.e:
```c++
if(project->isActiveConfig("lib_version_first"))
    project->values("TARGET_x").append(prefix + project->first("TARGET") + "." +
                                            project->first("VER_MAJ") + "." +
                                            project->first("QMAKE_EXTENSION_PLUGIN"));
```

**Comment by laanwj** [comment]
Code review ACK e1f2553e1148de9bd57818b40b5fe9da7ff5e246

**Pull request merged**